infrastructure/persistence/memo.go

```diff
@@ -0,0 +1,122 @@
+package persistence
+
+import (
+	"context"
+	"errors"
+	"github.com/Nexters/myply/domain/memos"
+	"github.com/Nexters/myply/infrastructure/persistence/db"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"time"
+)
+
+var collectionName = "memos"
+
+type MemoData struct {
+	ID             primitive.ObjectID   `json:"id" bson:"_id"`
+	DeviceToken    string               `json:"deviceToken"`
+	YoutubeVideoId string               `json:"youtubeVideoId"`
+	Body           string               `json:"body"`
+	TagIds         []primitive.ObjectID `json:"tagIds"`
+	CreatedAt      primitive.Timestamp  `json:"createdAt"`
+	UpdatedAt      primitive.Timestamp  `json:"updatedAt"`
+}
+
+type MemoMongoRepository struct {
+	conn *mongo.Database
+}
+
+func NewMemoRepository(i *db.MongoInstance) *memos.Repository {
+	var r memos.Repository
+	r = &MemoMongoRepository{conn: i.Db}
+	return &r
+}
+
+func (r *MemoMongoRepository) GetMemo(id string) (*memos.Memo, error) {
+	coll := r.getCollection()
+
+	objectId, err := primitive.ObjectIDFromHex(id)
+	if err != nil {
+		return nil, err
+	}
+
+	md := MemoData{}
+	if err = coll.FindOne(context.Background(), bson.M{"_id": objectId}).Decode(&md); err != nil {
+		switch {
+		case errors.Is(err, mongo.ErrNoDocuments):
+			return nil, memos.NotFoundException
+		default:
+			return nil, err
+		}
+	}
+
+	return r.toEntity(&md), nil
+}
+
+func (r *MemoMongoRepository) GetMemoByVideoId(id string) (*memos.Memo, error) {
+	coll := r.getCollection()
+
+	md := MemoData{}
+	if err := coll.FindOne(context.Background(), bson.M{"youtubeVideoId": id}).Decode(&md); err != nil {
+		switch {
+		case errors.Is(err, mongo.ErrNoDocuments):
+			return nil, memos.NotFoundException
+		default:
+			return nil, err
+		}
+	}
+
+	return r.toEntity(&md), nil
+}
+
+func (r *MemoMongoRepository) SaveMemo(videoId string, body string, deviceToken string) (string, error) {
+	coll := r.getCollection()
+
+	memoId := primitive.NewObjectID()
+	now := r.now()
+	md := MemoData{
+		ID:             memoId,
+		DeviceToken:    deviceToken,
+		YoutubeVideoId: videoId,
+		Body:           body,
+		TagIds:         nil, // TODO: change to real data
+		CreatedAt:      *now,
+		UpdatedAt:      *now,
+	}
+
+	insertResult, err := coll.InsertOne(context.Background(), md)
+	if err != nil {
+		return "", err
+	}
+
+	return insertResult.InsertedID.(primitive.ObjectID).Hex(), nil
+}
+
+func (r *MemoMongoRepository) getCollection() *mongo.Collection {
+	return r.conn.Collection(collectionName)
+}
+
+func (r *MemoMongoRepository) now() *primitive.Timestamp {
+	return &primitive.Timestamp{
+		T: uint32(time.Now().Unix()),
+		I: 0,
+	}
+}
+
+func (r *MemoMongoRepository) toEntity(m *MemoData) *memos.Memo {
```

Comment:
How about making it a method on `MemoData` rather than a method on it?

func (m *MemoData) toEntity() *memos.Memo {}

Reply:
i agree

Reply:
Avoid functions that return pointers.
Induce pressure on the gc.

```go
func (m MemoData) toEntity() memos.Memo {
   ...
}
```

Reply:
@minkj1992 Oh, it was a mistake. Thank you for letting me know :) I changed toEntity as a method for MemoData struct.

@mcauto Thanks! I just thought pass-by-reference would be better without profound grounds. Although I would like to deep dive gc issue right now, my PR should be merged first for the others to process other tasks. I would check this point as a left issue and refactor my codes soon :)